File: dgs-codegen/src/main/java/com/example/demo/gql/dataloaders/CommentsDataLoader.java
package com.example.demo.gql.dataloaders;

import com.example.demo.gql.types.Comment;
import com.example.demo.service.PostService;
import com.netflix.graphql.dgs.DgsDataLoader;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.dataloader.MappedBatchLoader;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

@DgsDataLoader(name = "comments")
@RequiredArgsConstructor
@Slf4j
public class CommentsDataLoader implements MappedBatchLoader<String, List<Comment>> {
    final PostService postService;

    @Override
    public CompletionStage<Map<String, List<Comment>>> load(Set<String> keys) {
        List<Comment> comments = this.postService.getCommentsByPostIdIn(keys);

        Map<String, List<Comment>> mappedComments = new HashMap<>();
        keys.forEach(
                k -> mappedComments.put(k, comments
                        .stream()
                        .filter(c -> c.getPostId().equals(k)).toList())
        );
        log.info("mapped comments: {}", mappedComments);
        return CompletableFuture.supplyAsync(() -> mappedComments);
    }
}

File: dgs-kotlin/src/main/kotlin/com/example/demo/gql/dataloaders/CommentsDataLoader.kt
package com.example.demo.gql.dataloaders

import com.example.demo.gql.types.Comment
import com.example.demo.service.PostService
import com.netflix.graphql.dgs.DgsDataLoader
import org.dataloader.MappedBatchLoader
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage

@DgsDataLoader(name = "comments")
class CommentsDataLoader(val postService: PostService) : MappedBatchLoader<UUID, List<Comment>> {

    override fun load(keys: Set<UUID>): CompletionStage<Map<UUID, List<Comment>>> {
        val comments = postService.getCommentsByPostIdIn(keys.toList())
        val mappedComments: MutableMap<UUID, List<Comment>> = hashMapOf()
        keys.forEach {
            mappedComments[it] = comments.filter { (_, _, postId) -> postId == it }
        }
        log.info("mapped comments: {}", mappedComments)
        return CompletableFuture.completedFuture(mappedComments)
    }

    companion object {
        val log: Logger = LoggerFactory.getLogger(CommentsDataLoader::class.java)
    }
}